README.md
Stub-O-Matic
============


Enable automated testing by mastering system dependencies.
Use when reality is simply not good enough.

Documentation
=============

See the `The documenation website <http://http://stubo-app.readthedocs.org/en/latest/>`_ to view
user documentation.

License
=======

Stub-O-Matic is offered under GPLv3, see LICENSE for more details.

Authors
=======

Stub-O-Matic is made available by `OpenCredo <http://opencredo.com>`_
and a team of contributors.


Install
=======

(Linux Red Hat Enterprise, CentOS, Fedora, or Amazon Linux)

Redis Server Installation::

    $ sudo yum install redis-server
    $ sudo service redis-server start

Mongo server Installation::

    $ sudo yum install mongodb-org
    $ sudo service mongod start | stop

dependencies::

    $ sudo yum install python27-virtualenv libxslt libxslt-devel libxml2 libxml2-devel texlive-full


(Linux Ubuntu) 

Redis Server Installation::

    $ sudo apt-get install -y redis-server
    $ sudo service redis-server start

Mongo server Installation::

    $ sudo apt-get install -y mongodb-org
    $ sudo service mongod start | stop
    
dependencies::

    $ sudo apt-get install -y python-dev python-pip wget python2.7-dev libxml2 libxml2-dev libxslt1-dev
    
Stubo::

    $ virtualenv --no-site-packages env
    $ source ./env/bin/activate
       
    (env) $ git clone git@github.com:Stub-O-Matic/stubo-app.git
    
    (env) $ cd stubo_app
    
    (env) $ python setup.py develop

RUN
===

Perform the following::

    $ sudo service mongod start
    $ sudo service redis-server start

Active your virtual python env:

    $ source ./env/bin/activate
    
You can start Stubo from your env root dir: 

    $ cd env
    
Note you need to create a log dir under the dir you call stubo from:
    
    (first run only)
    $ mkdir log
    $ create_tracker_collection
    
    $ stubo
    
Logging is in env/log/stubo.log
Default config is picked up from ./etc/dev.ini     
Alternatively you can run from another location:

    (env) $ stubo -c path_to_config


FUNCTIONAL TESTING
==================

You can run the test suite by running the following command 
(in the project root: stubo-app)::

        (env) $ nosetests stubo

Use the ``-x`` flag to stop the tests at the first error::

        (env) $ nosetests stubo -x
        
Or to run a specific test:

         (env) $ nosetests stubo.tests.integration.test_api:TestAcceptance.test_exec_first_commands

Use the ``--with-coverage`` flag to display a coverage report after
running the tests, this will show you which files / lines have not
been executed when the tests ran::

        (env) $ nosetests stubo --with-coverage --cover-package=stubo

Or to run a specific test:

        (env) $ nosetests stubo.tests.integration.test_api:TestAcceptance.test_exec_first_commands
        
and code analysis 

        (env) $ pylint stubo
        
DEBUGGING
=========

If you want to call functions to debug in a python shell and they have a 
dependency on redis or mongo you must initialise them first

    from stubo.utils import start_redis
    slave, master = start_redis({})
    slave
    <redis.client.Redis object at 0x10168e650> 
    
    from stubo.utils import init_mongo
    dbclient = init_mongo()
    dbclient
    Database(MongoClient('localhost', 27017), u'stubodb')

Now you can make calls that use redis

    >>> from stubo.cache import Cache
    >>> cache = Cache(host='localhost')
    >>> cache.get_session('first', 'first_1')
        {u'status': u'dormant', u'system_date': u'2015-02-24', u'scenario': u'localhost:first', u'last_used': u'2015-02-24 11:19:21', u'scenario_id': u'54ec5e3981875908f911a71b', u'session': u'first_1'}

& mongodb

    from stubo.model.db import Scenario
    scenario = Scenario()
    stubs = list(scenario.get_stubs('localhost:first'))
    import pprint
    pp = pprint.PrettyPrinter(indent=4)
    pp.pprint(stubs[0])
    {   u'_id': ObjectId('537c8f1cac5f7303ad704d85'),
        u'scenario': u'localhost:first',
        u'stub': {   u'recorded': u'2014-05-21',
                     u'request': {   u'bodyPatterns': [   {   u'contains': [   u'get my stub\n']}],
                                     u'method': u'POST'},
                     u'response': {   u'body': u'Hello {{1+1}} World\n',
                                      u'delayPolicy': u'slow',
                                      u'status': 200}}}
    
    from stubo.model.db import Tracker
    tracker = Tracker()
    list(tracker.find_tracker_data({'function' : 'get/response'}, 0, 1))
    [{u'function': u'get/response', u'scenario': u'conv', u'start_time': datetime.datetime(2013, 10, 29, 10, 38, 16, 534000, tzinfo=<bson.tz_util.FixedOffset object at 0x1015b6910>), u'return_code': 200, u'session': u'mary', u'duration_ms': 5, u'stubo_response': u'a2 response\n', u'_id': ObjectId('526f90180cfb4403fc27c0fa')}]
       
        
PERFORMANCE TESTING
===================
Load performance test data (over 10,000 stubs in 100 scenarios) with 

        stubo/api/exec/cmds?cmdfile=/static/cmds/perf/perf_setup.commands 
        this will take some time to run and leaves 100 sessions in playback mode.

        optionally demonstrate that the load was successfull with:
        stubo/api/exec/cmds?cmdfile=/static/cmds/perf/perf_getresponse.commands
        this gets one response from each session.

        When finished end sessions and delete stubs with:
        stubo/api/exec/cmds?cmdfile=/static/cmds/perf/perf_teardown.commands

ANALYTICS
=========
 
Certain Metrics are extracted from the track record and sent via UDP to a 
statsd server which aggregates the stats every 10 seconds before sending to 
a graphite carbon server via TCP. 

The location of the statsd server and metrics prefix to use
is specified in the stubo ini file

    statsd.host = localhost
    statsd.prefix = stubo
    
An analytics dashboard is provided to see that stats. Graphite connection 
details are specified in the stubo ini file::

    graphite.host = http://your-stubo-graphite.com/
    # auth to connect to graphite server if required 
    graphite.user = <user>
    graphite.password = <passwd>     

DOCUMENTATION
=============

Documentation is generated with Sphinx. The doc sources are located under /docs.

prereq for generating PDF:

    $ yum install texlive

To generate docs

    (env) $ python setup.py build_sphinx --build-dir=./stubo/static/docs

To generate PDF

       $ cd docs
       $ make latexpdf
       $ cp _build/latex/Stub-O-Matic.pdf ../stubo/static/docs

PROFILING
=========

Stubo profiling is available via two different profilers yappi and plop. 

yappi function stats

    name: name of the function being profiled
    #n: is the total callcount of the function.
    tsub: total spent time in the function minus the total time spent in the other functions called from this function.
    ttot: total time spent in the function.
    tavg: is same as ttot/ccnt. Average total time.

Plop is a stack-sampling profiler for Python. Profile collection can be turned on and off in a live process with minimal performance impact.

Both profilers are enabled via an HTTP inteface in stubo.

yappi - generate stats for 3 mins, output is saved to a csv file: 

    _profile?interval=180

plop - generate stats for 3 mins, output is saved by default to /tmp/plop.out:
    
    _profile2?interval=180&output=/tmp/plop.out

To view plop results in a bubble graph:

    (env) $ python -m plop.viewer --datadir=/tmp

View via the plop app at http://localhost:8888


CHANGES.txt
Stubo Changelog
=====================
0.7
----

0.6
----

Initial cut of open source version


testing.py
"""  
    :copyright: (c) 2015 by OpenCredo.
    :license: GPLv3, see LICENSE for more details.
"""
import os.path
import os
from collections import defaultdict
from tempfile import mkdtemp
from tornado.util import ObjectDict
from tornado.testing import AsyncHTTPTestCase
from tornado.web import MissingArgumentError

DummyModel = ObjectDict

def testdb_name():
    import uuid
    dbname = str(uuid.uuid4())
    return  'test_{0}'.format(dbname)
    
class Base(AsyncHTTPTestCase):
    
    def __init__(self, *args, **kwargs):
        self.cfg = kwargs.pop('cfg', {}) 
        super(Base, self).__init__(*args, **kwargs)  

    def setUp(self):
        super(Base, self).setUp()
        # the default settting of 5 secs is not enough for my old mac
        os.environ['ASYNC_TEST_TIMEOUT'] = '50'
        

    def tearDown(self):
        from stubo.utils import setup_redis
        self.redis_server.flushdb()  
        cache_db = setup_redis(db=9+1)
        cache_db.flushdb()
        self.db.connection.drop_database(self.testdb)
        self.app.settings['process_executor'].shutdown()
        super(Base, self).tearDown() 
              
    def get_app(self):
        from tornado.ioloop import IOLoop
        from stubo.service.run_stubo import TornadoManager
        from stubo.utils import init_mongo, start_redis, init_ext_cache
        self.testdb = testdb_name()

        self.cfg.update({
            'redis.host' : '127.0.0.1',
            'redis.port' : 6379,
            'redis.db' : 9,
            'redis_master.host' : '127.0.0.1',
            'redis_master.port' : 6379,
            'redis_master.db' : 9,
            'request_cache_limit' : 10,
        })  

        self.db = init_mongo({
            'tz_aware' : True,
            'db' : self.testdb
            })
        args = {'capped': True, 'size' : 100000}
        self.db.create_collection("tracker", **args)
        self.db.tracker.create_index('start_time', -1)

        # install() asserts that its not been initialised so setting it directly
        #self.io_loop.install() 
        IOLoop._instance = self.io_loop
        tm = TornadoManager(os.environ.get('STUBO_CONFIG_FILE_PATH'))
        self.redis_server, _ = start_redis(self.cfg)
        tm.cfg['ext_cache'] = init_ext_cache(self.cfg)
        tm.cfg['mongo.db'] = self.testdb
        tm.cfg.update(self.cfg)
        app = tm.get_app() 
        self.app = app
        from concurrent.futures import ProcessPoolExecutor
        self.app.settings['process_executor'] = ProcessPoolExecutor()
        return app     


         
class DummyRequestHandler(object):
    
    application_url = 'http://example.com'
    host = 'example.com:80'
    _ARG_DEFAULT = []

    def __init__(self, request=None, **kwargs):
        super(DummyRequestHandler, self).__init__()
        
        tmp_static_dir = mkdtemp()
        self.application = DummyModel(application_url=self.application_url,
                                      host=self.host,
                                      static_path=tmp_static_dir)
        self.application.update(**kwargs)  
        self.request = request or DummyModel(protocol='http', 
                                             host='localhost:8001',
                                             path='/',
                                             headers={},
                                             arguments=kwargs,
                                             body='')
        host, port = self.request.host.split(':')
        self.track = DummyModel(host=host, server=host, port=port, 
                                tracking_level='normal',
                                request_params=kwargs)

    def initialize(self):
        pass

    def prepare(self):
        """Called at the beginning of a request before `get`/`post`/etc."""
        self._start_time = time.time()

    def on_finish(self):
        """Called after the end of a request."""
        self._finish_time = time.time()
        
    def get_argument(self, name, default=_ARG_DEFAULT, strip=True):
        args = self.get_arguments(name, strip=strip)
        if not args:
            if default is self._ARG_DEFAULT:
                raise MissingArgumentError(name)
            return default
        return args[-1]   
    
    def get_arguments(self, name, strip=True):
        """Returns a list of the arguments with the given name.

        If the argument is not present, returns an empty list.

        The returned values are always unicode.
        """

        values = []
        for v in self.request.arguments.get(name, []):
            """"v = self.decode_argument(v, name=name)
            if isinstance(v, unicode_type):
                # Get rid of any weird control chars (unless decoding gave
                # us bytes, in which case leave it alone)
                v = RequestHandler._remove_control_chars_regex.sub(" ", v)"""
            if strip:
                v = v.strip()
            values.append(v)
        return values

    @property
    def settings(self):
        """ settings are stored directly in application."""
        return self.application

    def static_url(self, path):
        return os.path.join('/static', path)
    
    def request_time(self):
        """Returns the amount of time it took for this request to execute."""
        if self._finish_time is None:
            return time.time() - self._start_time
        else:
            return self._finish_time - self._start_time


class DummyHash(object):     
  
    def __init__(self, keys=None):
        self._keys = keys or {}
        
    def __call__(self, *args):
        return self    
    
    def keys(self, name):
        return self._keys.get(name, {}).keys()
    
    def remove(self, name):
        return self._keys.pop(name, None)
    
    def exists(self, key, name):
        item = self._keys.get(key)
        if not item:
            return False
        return name in item
    
    def delete(self, name, keys):
        item = self._keys.get(name)
        deleted = 0
        if item:
            if isinstance(keys, basestring):
                del item[keys]
                deleted += 1
            else:    
                for key in keys:
                    del item[key]
                    deleted += 1
            self._keys[name] = item
        return deleted         
    
    def get_raw(self, name, key):
        value = self._keys.get(name, {}).get(key)
        return value
      
    def get(self, name, key):
        import json
        try:
            return json.loads(self.get_raw(name, key))
        except TypeError:
            return None
    
    def get_all_raw(self, name):
        return self._keys.get(name, {})
        
    def get_all(self, name):
        import json
        result = self.get_all_raw(name)
        if result:
            result = dict((k, json.loads(v)) for k, v in result.iteritems())
        return result  
    
    def values(self, name):
        import json
        result = self.get_all_raw(name)
        if result:
            result = [json.loads(v) for v in result.itervalues()]
        return result             
    
    def set_raw(self, name, key, value):
        if name not in self._keys:
            self._keys[name] = {}    
        self._keys[name][key] = value 
        
    def set(self, name, key, value):
        import json
        self.set_raw(name, key, json.dumps(value)) 
        
    def incr(self, name, key, amount=1):
        val = self.get(name, key)
        if val:
            val += amount
        else:
            val = amount
        self.set(name, key, val)
        return val  
    
from stubo.cache import Cache
from stubo.model.db import Scenario
import ming
mg = ming.create_datastore('mim://')

class DummyScenario(Scenario):
    
    def __init__(self):
        dbname = testdb_name()
        client = getattr(mg.conn, dbname)
        Scenario.__init__(self, db=client)
        
    def __call__(self, **kwargs):
        return self  
    
    # derive these methods as ming hasn't implemented '$regex' matching
    
    def get_all(self, name=None):
        if name and isinstance(name, dict) and '$regex' in name:
            # {'$regex': 'localhost:.*'}
            name = name['$regex'].partition(':')[0]
            all_names =  list(Scenario.get_all(self))
            return [self.get(x.get('name')) for x in all_names if name in x.get('name')]
        else:
            return Scenario.get_all(self, name=name)  
        
        
    def get_stubs(self, name=None):
        if name and isinstance(name, dict) and '$regex' in name:
            name = name['$regex'].partition(':')[0]
            all_names =  list(self.db.scenario_stub.find({}))
            result =  [x for x in all_names if name in x.get('scenario')]
        else:
            result = Scenario.get_stubs(self, name=name)   
        return result       
    
class DummyCache(Cache):
    
    def __init__(self, host):
        Cache.__init__(self, host)
        self._hash =  DummyHash()
    
    def __call__(self, host):
        self.host = host
        return self    
    
    def hash_cls(self):
        return self._hash
    
    def get_all_saved_request_index_data(self):
        return {}
      


        
from collections import defaultdict
class DummyQueue(object):
    
    _data = defaultdict(list)
     
    def __init__(self, name, server=None):
        self.name = name
    
    def _get_list(self):
        return self._data[self.name]
    
    def get(self, timeout=None):
        return self._get_list().pop() 
    
    def get_item(self, index):
        return self._get_list()[index] if self._get_list() else None

    def put(self, msg):
        self._get_list().append(msg) 
        
    def delete(self):
        self._data.pop(self.name, 0)         
        
    def __len__(self):
        return len(self._get_list())           
        
class DummyTracker(object):
    
    def __init__(self, db=None):
        # db should be similar to a multidict
        self.db = db or defaultdict(list)
        
    def insert(self, track):
        _id = track.get('_id')
        if not _id:
            import time
            _id = int(time.time())          
        self.db[_id] = track
    
    def find_tracker_data(self, tracker_filter, skip, limit):
        self.db['_filter'] = tracker_filter
        self.db['_skip'] = skip
        self.db['_limit'] = limit
        return self.db

    def find_tracker_data_full(self, _id):
        return self.db.get(_id)
    
    def session_last_used(self, session):
        ''' Return the date this session was last used using the 
            last get/response time.
        '''
        import datetime
        return datetime.datetime.now()  
    
def make_stub(matchers, response, delay_policy=None, module=None,
              recorded=None):
    request = {
        "method": "POST",
        "bodyPatterns": [ { "contains": matchers } ]
    }
    response = {
        "status": 200,
        "body": response
    }
    if delay_policy:
        response['delayPolicy'] = delay_policy
    payload = {
        "recorded" : recorded or '2014-05-20',       
        "request": request,
        "response": response,
    }
    if module:
        payload['module'] = module
    return payload  
    
def make_cache_stub(matchers, response_ids, delay_policy=None, module=None,
                    recorded=None):
    stub = make_stub(matchers, response_ids, delay_policy=delay_policy, 
                     module=module, recorded=recorded)
    stub['response'].pop('body')
    stub['response']['ids'] = response_ids
    return stub
 